relay-server/src/http.rs

```diff
@@ -152,206 +129,118 @@ impl RequestBuilder {
-                    serde_json::from_slice(&bytes)
-                        .map_err(|e| HttpError::ActixJson(JsonPayloadError::Deserialize(e)))
-                });
-                Box::new(future) as Box<dyn Future<Item = _, Error = _>>
-            }
-        }
+        let future = self.bytes(limit).and_then(|bytes| {
+            serde_json::from_slice(&bytes)
+                .map_err(|e| HttpError::ActixJson(JsonPayloadError::Deserialize(e)))
+        });
+        Box::new(future) as Box<dyn Future<Item = _, Error = _>>
     }
 
     pub fn consume(mut self) -> ResponseFuture<Self, HttpError> {
-        // consume response bodies to allow connection keep-alive
-        match self {
-            Response::Actix(ref response) => Box::new(
-                response
-                    .payload()
-                    .for_each(|_| Ok(()))
-                    .map(|_| self)
-                    .map_err(HttpError::ActixPayload),
-            ),
-            Response::Reqwest(_) => {
-                // Consume the request payload such that the underlying connection returns to a
-                // "clean state".
-                //
-                // We do not understand if this is strictly necessary for reqwest. It was ported
-                // from actix-web where it was clearly necessary to un-break keepalive connections,
-                // but no testcase has been written for this and we are unsure on how to reproduce
-                // outside of prod. I (markus) have not found code in reqwest that would explicitly
-                // deal with this.
-                Box::new(
-                    // Note: The reqwest codepath is impossible to write with streams due to
-                    // borrowing issues. You *have* to use `chunk()`.
-                    async move {
-                        if let Response::Reqwest(ref mut response) = self {
-                            while response.chunk().await?.is_some() {}
-                        }
-                        Ok(self)
-                    }
-                    .boxed_local()
-                    .compat(),
-                )
+        // Consume the request payload such that the underlying connection returns to a
+        // "clean state".
+        //
+        // We do not understand if this is strictly necessary for reqwest. It was ported
+        // from actix-web where it was clearly necessary to un-break keepalive connections,
+        // but no testcase has been written for this and we are unsure on how to reproduce
+        // outside of prod. I (markus) have not found code in reqwest that would explicitly
+        // deal with this.
+        Box::new(
+            // Note: The reqwest codepath is impossible to write with streams due to
+            // borrowing issues. You *have* to use `chunk()`.
+            async move {
+                while self.0.chunk().await?.is_some() {}
+                Ok(self)
             }
-        }
+            .boxed_local()
```

Comment:
Is this `boxed_local` actually required if you box the compat future? Same below.

Reply:
you need to produce an Unpin type to call compat(), and the way to do so is to box the future. I don't really understand how that makes sense tbh but that is what the type constraints are.